apps/sync/js/fxsync-webcrypto/fxsyncwebcrypto.js

```diff
@@ -0,0 +1,279 @@
+'use strict';
+
+/* global StringConversion, KeyDerivation, crypto */
+/* exported FxSyncWebCrypto */
+
+// WebCrypto-based client for Firefox Sync.
+
+var FxSyncWebCrypto = (function fxsync_webcrypto() {
+
+  const HKDF_INFO_STR = 'identity.mozilla.com/picl/v1/oldsync';
+
+  // constructor
+  var FxSyncWebCrypto = function() {
+    // Basic check for presence of WebCrypto
+    if (!crypto || !crypto.subtle) {
+      throw new Error('This environment does not support WebCrypto');
+    }
+
+    this.mainSyncKey = null;
+    this.bulkKeyBundle = null;
+  };
+
+  function importKeyBundle(aesKeyAB, hmacKeyAB) {
+    var pAes = crypto.subtle.importKey('raw', aesKeyAB,
+                                          { name: 'AES-CBC', length: 256 },
+                                          true, [ 'encrypt', 'decrypt' ]
+                                    );
+    var pHmac =  crypto.subtle.importKey('raw', hmacKeyAB,
+                                          { name: 'HMAC', hash: 'SHA-256' },
+                                          true, [ 'sign', 'verify' ]
+                                      );
+    return Promise.all([pAes, pHmac]).then(function(results) {
+      return {
+        aes: results[0],
+        hmac: results[1]
+      };
+    });
+  }
+
+  FxSyncWebCrypto.prototype._importKb = function(kBByteArray) {
+    // The number 64 here comes from
+    // (256 bits for AES + 256 bits for HMAC) / (8 bits per byte)
+    return KeyDerivation.hkdf(kBByteArray,
+        StringConversion.rawStringToByteArray(HKDF_INFO_STR),
+        new Uint8Array(64), 64).then(function (output) {
+      var aesKeyAB = output.slice(0, 32).buffer;
+      var hmacKeyAB = output.slice(32).buffer;
+      return importKeyBundle(aesKeyAB, hmacKeyAB).then(function(keyBundle) {
+        this.mainSyncKey = keyBundle;
+      }.bind(this));
+    }.bind(this));
```

Comment:
No need to bind this if you use fat arrow functions. This applies for the whole patch.